### PR TITLE
Improve config center

### LINKED ISSUE #729: 配置中心功能完善
1、在dubbo-go/common/config/environment.go文件中，有如下注释，当前不支持“application level config center configuration ”这个是否有支持计划？
`// Environment
// There is dubbo.properties file and application level config center configuration which higner than normal config center in java. So in java the
// configuration sequence will be config center > application level config center > dubbo.properties > spring bean configuration.
// But in go, neither the dubbo.properties file or application level config center configuration will not support for the time being.
// We just have config center configuration which can override configuration in consumer.yaml & provider.yaml.
// But for add these features in future ,I finish the environment struct following Environment class in java.`

2、如下代码中配置中心的全局配置如果不存在，则直接返回报错，应用级的配置不去获取，后续是否会支持应用级配置。
`func (c *BaseConfig) prepareEnvironment() error {
	factory := extension.GetConfigCenterFactory(c.ConfigCenterConfig.Protocol)
	dynamicConfig, err := factory.GetDynamicConfiguration(c.configCenterUrl)
	config.GetEnvInstance().SetDynamicConfiguration(dynamicConfig)
	if err != nil {
		logger.Errorf("Get dynamic configuration error , error message is %v", err)
		return perrors.WithStack(err)
	}
	content, err := dynamicConfig.GetProperties(c.ConfigCenterConfig.ConfigFile, config_center.WithGroup(c.ConfigCenterConfig.Group))
	if err != nil {
		logger.Errorf("Get config content in dynamic configuration error , error message is %v", err)
		return perrors.WithStack(err)
	}`
3、如下应用级的配置会被全局配置覆盖掉
`func (env *Environment) Configuration() *list.List {
	cfgList := list.New()
	// The sequence would be: SystemConfiguration -> ExternalConfiguration -> AppExternalConfiguration -> AbstractConfig -> PropertiesConfiguration
	cfgList.PushFront(newInmemoryConfiguration(&(env.externalConfigMap)))
	cfgList.PushFront(newInmemoryConfiguration(&(env.appExternalConfigMap)))
	return cfgList
}

func (c *BaseConfig) fresh() {
	configList := config.GetEnvInstance().Configuration()
	for element := configList.Front(); element != nil; element = element.Next() {
		cfg := element.Value.(*config.InmemoryConfiguration)
		c.freshInternalConfig(cfg)
	}
}
`
综上，应该还是应用级的配置不支持，这个是否有开发计划？

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
**What this PR does**:
- Fix update of external config map
- AppExternalConfiguration orders before ExternalConfiguration
- Add config cache file
- Fix some typos

**Which issue(s) this PR fixes**:

Fixes #729 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by AlexStocks** [comment]
pls change the target branch to 1.5

**Comment by luckyxiaoqiang** [comment]
> pls change the target branch to 1.5

Updated.

File: common/config/environment.go [hunk 1]
```diff
@@ -72,22 +72,34 @@ func (env *Environment) UpdateExternalConfigMap(externalMap map[string]string) {
 	for k, v := range externalMap {
 		env.externalConfigMap.Store(k, v)
 	}
+	env.externalConfigMap.Range(func(key, value interface{}) bool {
+		if _, ok := externalMap[key.(string)]; !ok {
+			env.externalConfigMap.Delete(key)
```
Comment (flycash) [inline common/config/environment.go:77]: In which case this code will be excuted? 

It looks like that the unit tests do not cover this special case.

File: common/config/environment.go [hunk 2]
```diff
@@ -72,22 +72,34 @@ func (env *Environment) UpdateExternalConfigMap(externalMap map[string]string) {
...
 func (env *Environment) UpdateAppExternalConfigMap(externalMap map[string]string) {
 	for k, v := range externalMap {
 		env.appExternalConfigMap.Store(k, v)
 	}
+	env.appExternalConfigMap.Range(func(key, value interface{}) bool {
+		if _, ok := externalMap[key.(string)]; !ok {
+			env.appExternalConfigMap.Delete(key)
+		}
+		return true
+	})
```
Comment (flycash) [inline common/config/environment.go:93]: I think those lines could be refactor and extract them to be a new function

File: config/config_loader.go [hunk 3]
```diff
@@ -138,6 +140,17 @@ func loadConsumerConfig() {
...
+	// Write current configuration to cache file.
+	if consumerConfig.CacheFile != "" {
+		if data, err := yaml.MarshalYML(consumerConfig); err != nil {
+			logger.Errorf("Marshal consumer config err: %s", err.Error())
+		} else {
+			if err := ioutil.WriteFile(consumerConfig.CacheFile, data, 0666); err != nil {
+				logger.Errorf("Write consumer config cache file err: %s", err.Error())
+			}
+		}
+	}
```
Comment (flycash) [inline config/config_loader.go:152]: I am not sure whether we should do this like:
```go
(c *consumerConfig) SaveToCacheFile() {
    // ...
}

// in this funtion

consumerConfig.SaveToFile()
```

File: config/config_loader.go [hunk 4]
```diff
@@ -194,6 +207,17 @@ func loadProviderConfig() {
...
+	// Write the current configuration to cache file.
+	if providerConfig.CacheFile != "" {
+		if data, err := yaml.MarshalYML(providerConfig); err != nil {
+			logger.Errorf("Marshal provider config err: %s", err.Error())
+		} else {
+			if err := ioutil.WriteFile(providerConfig.CacheFile, data, 0666); err != nil {
+				logger.Errorf("Write provider config cache file err: %s", err.Error())
+			}
+		}
+	}
```
Comment (flycash) [inline config/config_loader.go:219]: See my above comment

File: config/config_loader.go [hunk 5]
```diff
@@ -138,6 +140,17 @@ func loadConsumerConfig() {
 		ref.Implement(rpcService)
 	}
 
+	// Write current configuration to cache file.
+	if consumerConfig.CacheFile != "" {
+		if data, err := yaml.MarshalYML(consumerConfig); err != nil {
+			logger.Errorf("Marshal consumer config err: %s", err.Error())
+		} else {
+			if err := ioutil.WriteFile(consumerConfig.CacheFile, data, 0666); err != nil {
```
Comment (zouyx) [inline config/config_loader.go:148]: I can not get this point. why will you copy this config to a new file?

Comment (luckyxiaoqiang) [reply config/config_loader.go:148]: Please see the issue comment: https://github.com/apache/dubbo-go/issues/729#issuecomment-681617545.

File: common/config/environment.go (hunk 1 above)
Comment (luckyxiaoqiang) [reply common/config/environment.go:77]: It will be executed when a key of external config deleted.

Comment (luckyxiaoqiang) [reply common/config/environment.go:77]: Units tests already covered, please see the change of test file.

**Review by AlexStocks — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Review by flycash — APPROVED** [review]

File: common/config/environment.go (hunk 1 above)
Comment (LaurenceLiZhixin) [reply common/config/environment.go:77]: Why not clear old map and copy from externalMap once.

File: config/base_config.go [hunk 6]
```diff
@@ -50,6 +50,9 @@ type BaseConfig struct {
 	EventDispatcherType string        `default:"direct" yaml:"event_dispatcher_type" json:"event_dispatcher_type,omitempty"`
 	MetricConfig        *MetricConfig `yaml:"metrics" json:"metrics,omitempty"`
 	fileStream          *bytes.Buffer
+
+	// cache file used to store the current used configurations.
+	CacheFile string `yaml:"cache_file" json:"cache_file,omitempty" property:"cache_file"`
```
Comment (LaurenceLiZhixin) [inline config/base_config.go:55]: I can't find where to use this field.

Comment (luckyxiaoqiang) [reply config/base_config.go:55]: Please see `config_loader.go` file.

Comment (LaurenceLiZhixin) [reply config/base_config.go:55]: Yes, you write provider config to CacheFile, but I want to know where to use it.

Comment (luckyxiaoqiang) [reply config/base_config.go:55]: Please see the issue comment: #729 (comment).

**Review by LaurenceLiZhixin — APPROVED** [review]

**Pull request merged**